README.md
R interface to X-13ARIMA-SEATS
------------------------------

<!-- badges: start -->
[![R-CMD-check](https://github.com/christophsax/seasonal/workflows/R-CMD-check/badge.svg)](https://github.com/christophsax/seasonal/actions)
[![Downloads](http://cranlogs.r-pkg.org/badges/seasonal)](https://cran.r-project.org/package=seasonal)
<!-- [![Codecov test coverage](https://codecov.io/gh/christophsax/seasonal/branch/main/graph/badge.svg)](https://codecov.io/gh/christophsax/seasonal?branch=main) -->
<!-- badges: end -->




*seasonal* is an easy-to-use and full-featured R-interface to X-13ARIMA-SEATS,
the newest seasonal adjustment software developed by the [United States Census
Bureau][census].

<img src="https://raw.githubusercontent.com/christophsax/seasonal/master/vignettes/Figures/website.jpg" width="65%"/>

### Installation

*seasonal* depends on the [x13binary][x13binary] package to access pre-built
binaries of X-13ARIMA-SEATS on all platforms and does not require any manual
installation. To install both packages:

    install.packages("seasonal")

### Getting started

`seas` is the core function of the *seasonal* package. By default, `seas` calls
the automatic procedures of X-13ARIMA-SEATS to perform a seasonal adjustment
that works well in most circumstances:

    m <- seas(AirPassengers)

For a more detailed introduction, check our [article in the *Journal of
Statistical Software*](https://doi.org/10.18637/jss.v087.i11) or consider the
vignette:

    vignette("seas")

### Input

In *seasonal*, it is possible to use almost the complete syntax of
X-13ARIMA-SEATS. The X-13ARIMA-SEATS syntax uses *specs* and *arguments*, with each spec
optionally containing some arguments. These spec-argument combinations can be
added to `seas` by separating the spec and the argument by a dot (`.`). For
example, in order to set the 'variables' argument of the 'regression' spec equal
to `td` and `ao1999.jan`, the input to `seas` looks like this:

    m <- seas(AirPassengers, regression.variables = c("td", "ao1955.jan"))

The best way to learn about the relationship between the syntax of
X-13ARIMA-SEATS and *seasonal* is to study the [comprehensive list of examples][examples].
Detailed information on the options can be found in the Census Bureaus'
[official manual][manual].

### Output

*seasonal* has a flexible mechanism to read data from X-13ARIMA-SEATS. With the
`series` function, it is possible to import almost all output that can be
generated by X-13ARIMA-SEATS. For example, the following command returns the
forecasts of the ARIMA model as a `"ts"` time series:

    m <- seas(AirPassengers)
    series(m, "forecast.forecasts")

### Graphs

There are several graphical tools to analyze a `seas` model. The main plot
function draws the seasonally adjusted and unadjusted series, as well as the
outliers:

    m <- seas(AirPassengers, regression.aictest = c("td", "easter"))
    plot(m)


### Graphical User Interface

The `view` function is a graphical tool for choosing a seasonal adjustment
model, using the [seasonalview][seasonalview] package, with the same
structure as the [demo website of seasonal][seasweb]. To install
[seasonalview][seasonalview], type:

    install.packages("seasonalview")

The goal of `view` is to summarize all relevant options, plots and statistics
that should be usually considered. `view` uses a `"seas"` object as its main
argument:

    view(m)

### License

*seasonal* is free and open source, licensed under GPL-3. It requires the
X-13ARIMA-SEATS software by the U.S. Census Bureau, which is open source and
freely available under the terms of its own [license][license].

To cite seasonal in publications use:

Sax C, Eddelbuettel D (2018). “Seasonal Adjustment by X-13ARIMA-SEATS
in R.” _Journal of Statistical Software_, *87*(11), 1-17. doi:
10.18637/jss.v087.i11 (URL: https://doi.org/10.18637/jss.v087.i11).

Please report bugs and suggestions on [GitHub][github]. Thank you!

[seasonalview]: https://cran.r-project.org/package=seasonalview "Graphical User Interface for Seasonal Adjustment"

[x13binary]: https://cran.r-project.org/package=x13binary "X-13ARIMA-SEATS binary for R"

[manual]: https://www2.census.gov/software/x-13arima-seats/x13as/windows/documentation/docx13as.pdf "Reference Manual"

[census]: https://www.census.gov/data/software/x13as.X-13ARIMA-SEATS.html "United States Census Bureau"

[license]: https://www.census.gov/data/software/x13as/disclaimer.html "License Information and Disclaimer"

[examples]: http://www.seasonal.website/examples.html "Examples of X-13ARIMA-SEATS in R"

[seasweb]: http://www.seasonal.website "Examples of X-13ARIMA-SEATS in R"

[github]: https://github.com/christophsax/seasonal "Development on Github"


NEWS
1.8.4.3 (unreleased)
--------------------------------------------------------------------------------
  - seas() works within forking parallization (e.g., mclapply) #276
    Thanks to @jonlachmann!
  - annual series are read correctly #264
  - Jan 1 can be used in genhol() #261
  - genhol() produces correct result in 1931 #193


1.8.4.2 (unreleased)
--------------------------------------------------------------------------------
  - out() works on composite objects #278
  - series() works on composite objects #278


1.8.4.1 (unreleased)
--------------------------------------------------------------------------------
  - Keep series specification also in `noinst/specs/specs.csv`
  - Adjust specs to include composite series, some additional seats series


1.8.4 2021-09-17
--------------------------------------------------------------------------------

bug fix
  - Avoid setting X13_PATH environment variable. Fix M1 startup issue #267


1.8.3  2021-06-14
--------------------------------------------------------------------------------

bug fix
  - Make sure example(seas) does not leave anything behind
  - escape tests on M1


1.8.2  2021-03-23
--------------------------------------------------------------------------------

bug fix
  - CRAN M1: rm detritus in the temp directory


1.8.1  2021-03-13
--------------------------------------------------------------------------------

bug fix
  - CRAN: do not run example on Solaris


1.8.0  2021-03-05
--------------------------------------------------------------------------------

new features
  - support for multiple time series and the composite spec

enhancements
  - static() prints a prettier output

bug fix
  - as.data.frame() fix date conversion for freqency 4, 2 (#256)
  - static() drop occasional warning (#235)
  - fix some backslash confusion on Windows (#237)


1.7.1  2020-06-06
--------------------------------------------------------------------------------

bug fix
  - specify convert.type(as.is = TRUE), to comply with next R


1.7.0  2018-12-19
--------------------------------------------------------------------------------

minor changes

  - like update(), static now uses m$list, rather than m$call. This makes it
    usable in lapply or other functions.

bug fix
  - fix dates for Chinese New Year in 2015 and 2017 (#219)
  - outlier(): reverted workaround for problem with byte compiler in R 3.4.0,
    which has been fixed in R. Fixes incomplete outliers in view() and bug
    in identify. (#211, #212)
  - fix error in summary when no seasonal adjustment is requested (#226)

documentation
  - article in Journal of Statistical Software as vignette
  - README.md providing a short overview of the package
  - citation refering to publication in Journal of Statistical Software


1.6.1  2017-05-02
--------------------------------------------------------------------------------

enhancements

  - A second, more verbose run after a non-zero exit, to detect Fortran run time
    errors (https://github.com/christophsax/seasonal/issues/194)

bug fix
  - deal with new file in build 1.1.39 (macOS)
  - outlier(): workaround for problem with byte compiler in R 3.4.0
    https://github.com/christophsax/seasonal/issues/209
  - import.spc(): fixed parsing of explicitly specified NULL entries
    E.g., regression{aictest=()} results in now regression.aictest = NULL
    (https://github.com/christophsax/seasonal/issues/202)
  - series(), out(): fix if model was updated using update()
    (https://github.com/christophsax/seasonal/issues/191)
  - some X-13 notes were not shown in summary
    (https://github.com/christophsax/seasonal/issues/200)
  - static parsing error with change of regime
    (https://github.com/christophsax/seasonal/issues/205)


1.5.1  2017-02-12
--------------------------------------------------------------------------------

bug fix
  - removing 'udg' classification and as.data.frame() method from udg() output,
    because it messes up with several other functions.


1.5.0  2017-02-12
--------------------------------------------------------------------------------

defunct
  - inspect() is defunct, use view() instead. Shiny application code removed
    from package.

enhancements
  - as.data.frame() methods to coerce output of seas, summary.seas() and udg()
    to a data.frame. Useful for further processing (experimental). Replaces the
    experimental 'data.frame' argument in summary.seas() and udg() from the last
    version.


1.4.0  2016-12-23
--------------------------------------------------------------------------------

new features
  - seasonal now uses the graphical user interface from the new seaonalview
    package. To install from CRAN, use: install.packages('seasonalview')
    Use the 'view()' function instead of 'inspect', which is deprecated and will
    be defunct soon.

enhancements
  - new update() method, updates a "seas" object, works with R generic.
  - new predict() method, returns the final series of an (updated) "seas"
    object, works with R generic.
  - static() function has new argument 'x11.static'. If TRUE, moving averages
    of X-11 are fixed as well.
    (Thanks to Brian Monsell)
  - static() function has new argument 'evaluate'. If TRUE, it returns an
    evaluated call, an object of class "seas".
  - import.spc() function has new argument 'text', to accept the content of an
    X-13 .spc file as a character vector.
  - udg() and summary.seas() gain an optional 'data.frame' argument, which
    returns their info as a data.frame. Useful for further processing.
    (experimental)

bug fixes
  - import.spc imports arima models with commas. E.g., (0, 1, 1).
    (Thanks to Brian Monsell)
  - substitute colons in series name to avoid reading error.
  - shorten long series names to avoid X-13 error.
    (Both thanks to Bill Dunlap)
  - missing udg element in summary no longer leads to an error.
  - xreg specification ensures regression spec is used.
    (Thanks to Jose Pavia)

R/seas.R
#' Seasonal Adjustment with X-13ARIMA-SEATS
#'
#' Main function of the seasonal package. With the default options,
#' `seas` calls the  automatic procedures of X-13ARIMA-SEATS to perform a
#' seasonal adjustment that works well in most circumstances. Via the `...`
#' argument, it is possible to invoke almost all options that are available in
#' X-13ARIMA-SEATS (see  details). The default options of `seas` are listed
#' as explicit arguments and are  discussed in the arguments section. A
#' full-featured graphical user interface can be accessed by the
#' [view()] function.
#'
#'  It is possible to use the almost complete syntax of X-13ARIMA-SEAT via the
#' `...` argument. The syntax of X-13ARIMA-SEATS uses *specs* and
#' *arguments*, and each spec optionally contains some arguments. In
#' `seas`, an additional spec-argument can be added by separating spec and
#' argument by a dot (`.`) (see examples). Alternatively, spec-argument
#' combinations can be supplied as a named list, which is useful for
#' programming.
#'
#' Similarly, the [series()] function can be used to read almost all
#' series from X-13ARIMA-SEATS. The [udg()] function provides access
#' to a large number of diagnostical statistics.
#'
#' For a more extensive description, consider `vignette("seas")` or the wiki page,
#' which contains replications of almost all examples from the official
#' X-13ARIMA-SEATS manual.
#'
#' @param x   object of class `"ts"` or `"mts", or a list of `"ts"` objects: time series to seasonally adjust.
#' @param xreg   (optional) object of class `"ts"`: one or several user
#'   defined exogenous variables for regARIMA modeling, can be used both with
#'   `regression` or `x11regression`.
#' @param xtrans   (optional) object of class `"ts"`: one or two user
#'   defined exogenous variables for the `transform` spec. Can be specifed
#'   together with `xreg`.
#' @param seats.noadmiss   spec 'seats' with argument `noadmiss = "yes"`
#'   (default). Seasonal adjustment by SEATS, if SEATS decomposition is invalid,
#'   an alternative model is used (a message is returned). If `noadmiss =
#'   "no"`, no approximation is done. If the seats spec is removed
#'   (`seats = NULL`), no seasonal adjustment is performed.
#' @param transform.function   spec `transform` with argument
#'   `function = "auto"` (default). Automatic log transformation detection.
#'   Set equal to `"none"`, `"log"` or any value that is allowed by
#'   X-13 to turn it off.
#' @param regression.aictest   spec `regression` with argument
#'   `aictest = c("td", "easter")` (default). AIC test for trading days and
#'   Easter effects. Set equal to `NULL` to turn it off.
#' @param outlier   spec `outlier` without arguments (default). Automatic
#'   outlier detection. Set equal to `NULL` to turn it off.
#' @param automdl   spec `automdl` without arguments (default). Automatic
#'   model search with the automdl spec. Set equal to `NULL` to turn it off.
#' @param composite   spec `composite`. A named list with spec-arguments
#'   for the aggregation of multiple series. Also requries
#'   `series.comtype = "add"` or similar. Set equal to `NULL` to turn
#'   it off (default). See `vignette("multiple")`.
#' @param na.action  a function which indicates what should happen when the data
#'   contain NAs. `na.omit` (default), `na.exclude` or `na.fail`.
#'   If `na.action = na.x13`, NA handling is done by X-13, i.e. NA values
#'   are substituted by -99999.
#' @param out   logical. Should the X-13ARIMA-SEATS standard output be saved in
#'   the `"seas"` object? (this increases object size substantially, it is
#'   recommended to re-evaluate the model using the [out()] function
#'   instead.)
#' @param dir   character string with a user defined file path. If specified,
#'   the X-13ARIMA-SEATS output files are copied to this folder. Useful for
#'   debugging.
#' @param multimode   one of `"x13"` or `"R"`. When multiple series
#'   are supplied, should they be processed in a single call (`"x13"`) or
#'   processed individually (`"R"`). See `vignette("multiple")`.
#' @param ...  additional spec-arguments options sent to X-13ARIMA-SEATS (see
#'   details).
#' @param list  a named list with additional spec-arguments options. This is an
#'   alternative to the `...` argument. It is useful for programming.
#'
#' @return returns an object of class `"seas"`, essentially a list with the
#'   following components:
#'   \item{series}{a list containing the output tables of X-13. To be accessed
#'   by the `series` function.}
#'   \item{data}{seasonally adjusted data, the
#'   raw data, the trend component, the irregular component and the seasonal
#'   component (deprecated).}
#'   \item{err}{warning messages from X-13ARIMA-SEATS}
#'   \item{udg}{content of the `.udg` output file}
#'   \item{est}{content of the `.est` output file}
#'   \item{model}{list with the model specification,
#'   similar to `"spc"`. It typically contains `"regression"`, which
#'   contains the regressors and parameter estimates, and `"arima"`, which
#'   contains the ARIMA specification and the parameter estimates.}
#'   \item{fivebestmdl}{Best Five ARIMA Models (unparsed)}
#'   \item{x}{input series}
#'   \item{spc}{object of class `"spclist"`, a list containing the content of the `.spc` file that is
#'   used by X-13ARIMA-SEATS. Each spec is on the first level, each
#'   argument is on the second level.}
#'   \item{call}{function call}
#'   \item{wdir}{temporary directory in which X-13ARIMA-SEATS has been run}
#'
#'   The `final` function returns the final adjusted series, the
#'   `plot` method shows a plot with the unadjusted and the adjusted
#'   series. `summary` gives an overview of the regARIMA model. The
#'   [udg()] function returns diagnostical statistics.
#'
#' @seealso [view()], for accessing the graphical user interface.
#' @seealso [update.seas()], to update an existing `"seas"`
#'   model.
#' @seealso [static()], to return the 'static' call, with automated
#'   procedures substituted by their choices.
#' @seealso [series()], for universal X-13 table series import.
#' @seealso [out()], to view the full X-13 diagnostical output.
#'
#' @references
#'   Sax C, Eddelbuettel D (2018). "Seasonal Adjustment by X-13ARIMA-SEATS
#'   in R." *Journal of Statistical Software*, *87*(11), 1-17.
#'   \doi{10.18637/jss.v087.i11}.
#'
#'   On-Line Interface to seasonal
#'   <http://www.seasonal.website>
#'
#'   Comprehensive list of R examples from the X-13ARIMA-SEATS manual:
#'   <http://www.seasonal.website/examples.html>
#'
#'   Official X-13ARIMA-SEATS manual:
#'   <https://www2.census.gov/software/x-13arima-seats/x13as/windows/documentation/docx13as.pdf>
#' @export
#' @import datasets
#' @import grDevices
#' @import graphics
#' @import utils
#'
#' @examples
#'
#' # Basic call
#' \donttest{
#' m <- seas(AirPassengers)
#' summary(m)
#' }
#'
#' # Graphical user interface
#' \dontrun{
#' view(m)
#' }
#'
#' \donttest{
#' # invoke X-13ARIMA-SEATS options as 'spec.argument' through the ... argument
#' # (consult the X-13ARIMA-SEATS manual for many more options and the list of
#' # R examples for more examples)
#' seas(AirPassengers, regression.aictest = c("td"))  # no easter testing
#' seas(AirPassengers, force.type = "denton")  # force equality of annual values
#' seas(AirPassengers, x11 = "")  # use x11, overrides the 'seats' spec
#'
#' # 'spec.argument' combinations can also be supplied as a named list, which is
#' # useful for programming
#' seas(AirPassengers, list = list(regression.aictest = c("td"), outlier = NULL))
#' # constructing the list step by step
#' ll <- list()
#' ll[["x"]] <- AirPassengers
#' ll[["regression.aictest"]] <- "td"
#' ll["outlier"] <- list(NULL)  # assigning NULL to a list using single brackets
#' seas(list = ll)
#'
#' # options can be entered as vectors
#' seas(AirPassengers, regression.variables = c("td1coef", "easter[1]"))
#' seas(AirPassengers, arima.model = c(0, 1, 1, 0, 1, 1))
#' seas(AirPassengers, arima.model = "(0 1 1)(0 1 1)")     # equivalent
#'
#' # turn off the automatic procedures
#' seas(AirPassengers, regression.variables = c("td1coef", "easter[1]",
#' "ao1951.May"), arima.model = "(0 1 1)(0 1 1)", regression.aictest = NULL,
#' outlier = NULL, transform.function = "log")
#'
#' # static replication of 'm <- seas(AirPassengers)'
#' static(m)  # this also tests the equivalence of the static call
#' static(m, test = FALSE)  # no testing (much faster)
#' static(m, coef = TRUE)  # also fixes the coefficients
#'
#' # updating an existing model
#' update(m, x11 = "")
#'
#' # specific extractor functions
#' final(m)
#' predict(m)   # equivalent
#' original(m)
#' resid(m)
#' coef(m)
#' fivebestmdl(m)
#' spc(m)                  # the .spc input file to X-13 (for debugging)
#'
#' # universal extractor function for any X-13ARIMA-SEATS output (see ?series)
#' series(m, "forecast.forecasts")
#'
#' # user defined regressors (see ?genhol for more examples)
#' # a temporary level shift in R base
#' tls <- ts(0, start = 1949, end = 1965, freq = 12)
#' window(tls, start = c(1955, 1), end = c(1957, 12)) <- 1
#' seas(AirPassengers, xreg = tls, outlier = NULL)
#' # identical to a X-13ARIMA-SEATS specification of the the level shift
#' seas(AirPassengers, regression.variables = c("tl1955.01-1957.12"),
#'      outlier = NULL)
#'
#' # forecasting an annual series without seasonal adjustment
#' m <- seas(airmiles, seats = NULL, regression.aictest = NULL)
#' series(m, "forecast.forecasts")
#'
#' # NA handling
#' AirPassengersNA <- window(AirPassengers, end = 1962, extend = TRUE)
#' final(seas(AirPassengersNA, na.action = na.omit))    # no NA in final series
#' final(seas(AirPassengersNA, na.action = na.exclude)) # NA in final series
#' # final(seas(AirPassengersNA, na.action = na.fail))    # fails
#'
#' # NA handling by X-13 (works with internal NAs)
#' AirPassengersNA[20] <- NA
#' final(seas(AirPassengersNA, na.action = na.x13))
#'
#' ## performing 'composite' adjustment
#' seas(
#'   cbind(mdeaths, fdeaths),
#'   composite = list(),
#'   series.comptype = "add"
#' )
#' }
seas <- function(x = NULL, xreg = NULL, xtrans = NULL,
         seats.noadmiss = "yes", transform.function = "auto",
         regression.aictest = c("td", "easter"), outlier = "",
         automdl = "", composite = NULL, na.action = na.omit,
         out = FALSE, dir = NULL, multimode = c("x13", "R"), ..., list = NULL){

  # initial checks
  checkX13(fail = TRUE, fullcheck = FALSE, htmlcheck = FALSE)


  # multi mode
  multi_x <- (!is.null(x) && NCOL(x) > 1) || is.list(x)
  multi_list <-
    !is.null(list) &&
    all(sapply(list, inherits, "list")) &&
    length(list) > 1

  if (multi_x || multi_list) {
    z <- seas_multi(
      x = x,
      xreg = xreg,
      xtrans = xtrans,
      seats.noadmiss = seats.noadmiss,
      transform.function = transform.function,
      regression.aictest = regression.aictest,
      outlier = outlier,
      automdl = automdl,
      composite = composite,
      na.action = na.action,
      out = out,
      dir = dir,
      list_dots = list(...),
      multimode = multimode,
      list = list,
      call = match.call()
    )
    return(z)
  }

  # standard run
  list_combined <- enrich_list(
    list = list,
    list_dots = list(...),
    x = x,
    xreg = xreg,
    xtrans = xtrans,
    seats.noadmiss = seats.noadmiss,
    transform.function = transform.function,
    regression.aictest = regression.aictest,
    outlier = outlier,
    automdl = automdl
  )

  seas_list(
    list = list_combined,
    na.action = na.action,
    out = out,
    dir = dir,
    call = match.call(),
    series.name = deparse(substitute(x))[1]
  )
}
